components/microsoft_onedrive/microsoft_onedrive.app.mjs

```diff
@@ -0,0 +1,268 @@
+          },
+        };
+      },
+    },
+    fileTypes: {
+      type: "string[]",
+      label: "File Types",
+      description: "The types of files to watch for",
+      options: mimeTypes,
+    },
+  },
+  methods: {
+    /**
+     * @returns Microsoft Graph Client instance
+     */
+    client() {
+      return new Client.initWithMiddleware({
+        authProvider: {
+          getAccessToken: () => Promise.resolve(this.$auth.oauth_access_token),
+        },
+      });
+    },
+    /**
+     * This is a utility method that returns the path to a OneDrive drive based
+     * on its identifier, or the authenticated user's drive if an identifier is
+     * not specified.
+     *
+     * @param {string}  [driveId] the OneDrive drive identifier. When not
+     * provided, the method returns the path of the authenticated user's drive.
+     * @returns the path to the specified drive
+     */
+    _getDrivePath(driveId) {
+      return driveId
+        ? `/drives/${driveId}`
+        : "/me/drive";
+    },
+    /**
+     * This is a utility method that returns the path to a OneDrive item based
+     * on its identifier, or the root if an identifier is not specified.
+     *
+     * @param {string}  [driveItemId] the OneDrive item identifier. When not
+     * provided, the method returns the path of the root item.
+     * @returns the path to the specified drive
+     */
+    _getDriveItemPath(driveItemId) {
+      return driveItemId
+        ? `/items/${driveItemId}`
+        : "/root";
+    },
+    _subscriptionsEndpoint(id) {
+      return id
+        ? `/subscriptions/${id}`
+        : "/subscriptions";
+    },
+    _getDefaultHookExpirationDateTime() {
+      // 30 days from now
+      const futureTimestamp = Date.now() + 43200 * 60 * 1000;
+      return new Date(futureTimestamp).toISOString();
+    },
+    /**
+     * This method creates a [OneDrive webhook](https://bit.ly/2PxfQ9j) to
+     * monitor a specific resource, defaulted to the authenticated user's drive.
+     *
+     * @param {string}  notificationUrl the target URL of the webhook
+     * @param {object}  [opts] an object containing the different options for
+     * the hook creation
+     * @param {string}  [opts.driveId] the resource to which the webhook will
+     * subscribe for events
+     * @param {string}  [opts.expirationDateTime] the timestamp of the hook
+     * subscription expiration, in ISO-8601 format. Defaults to 30 days after
+     * the time this method is called.
+     * @returns the ID of the created webhook
+     */
+    async createHook(notificationUrl, {
+      driveId,
+      expirationDateTime = this._getDefaultHookExpirationDateTime(),
+    } = {}) {
+      return new Promise((resolve, reject) => {
+        this.client()
+          .api(this._subscriptionsEndpoint())
+          .post({
+            notificationUrl,
+            resource: this._getDrivePath(driveId) + "/root",
+            expirationDateTime,
+            changeType: "updated",
```

Comment:
Can the `resource` here be set to only watch a specific folder instead of the whole drive (maybe using `driveItemPath`)? That would reduce unnecessary invocations for the components that watch a folder.

Reply:
I did some tests and I was not able to do it. But as far as I know, the One drive sends an event and we need to search the changes using a `deltaLink`. In this `deltaLink` you can specify some things such as the `folder`. This is already implemented

Reply:
I did some tests and it's not possible. But the hook has a `deltaLink` and we can pass some configurations such as `folderId`.